### PR TITLE
Workbench Grid Layout Behind flag

### PR DESCRIPTION
**Pull request by sbatten**
partially functioning grid view

There is so much CSS using .monaco-workbench > .part that needed to be changed. It is currently changed to just remove the child selector but this can make css rule lookup slow and I think it would be better to just have rules that apply to parts not need the .monaco-workbench parent

- [x] There are compile errors in the branch. I lost some time rebuilding a couple times, thinking it was my fault.
- [x] The panel/viewlet views don't get removed when they are collapsed.
- [x] The sidebar doesn't snap yet.
- [x] Toggle side bar position is broken.
- [x] Hide Activity Bar is broken.
- [x] Toggle titlebar with fullscreen and menu toggling
- [ ] Panel still grows proportionally
- [ ] Zen mode is broken.

### CODE REVIEW

**Comment by joaomoreno** [comment]
Holy cow! Great job, though. I also want to get @bpasero's eyes on this, since he owns pretty much all of it.

**Comment by beastdestroyer** [comment]
Is the grid layout going to apply everywhere? Including UIs like the welcome page, panels etc?

**Comment by joaomoreno** [comment]
@beastdestroyer Please read https://github.com/Microsoft/vscode/issues/50853#issuecomment-438606737:

> This is very likely not what you think it is. Though we appreciate your passion.
> 
> This issue aims to use the grid widget for layouting. It will not introduce any new layouting features. It will not enable you to move viewlets around. It is a purely engineering issue to clean up some old layouting code.

**Comment by joaomoreno** [comment]
@sbatten I was able to test this. Some comments:

- There are compile errors in the branch. I lost some time rebuilding a couple times, thinking it was my fault.
- The panel/viewlet views don't get removed when they are collapsed.
- The sidebar doesn't snap yet.
- Views still seem to grow proportionally instead of using the resizing priority feature.
- Toggle side bar position is broken.
- Hide Activity Bar is broken.
- Zen mode is broken.

Feel free to pick these up into TODO check boxes for this issue.

Still, it's very exciting seeing it all come together. Considering the immense effort this is, great job! :clap:

**Comment by bpasero** [comment]
@sbatten cool stuff. yeah I agree it would probably be fine to get rid of `monaco-workbench` to speed up CSS selector matching as long as we can make sure the selector is unique.

`layout.ts` seems to have compile errors but I assume this file can be deleted anyways.

On macOS I see the titlebar part with space for custom menu showing up even though we never use custom menu on macOS. But since this is WIP, I will not do a review yet 👍

**Comment by sbatten** [comment]
@joaomoreno maybe the snapSize value should be a number from 0 to 1 in relation to the minimum size of the current orientation. panel needs to have a different snap value depending on panel position. also, it seems that if an item is snapped closed, that state isn't serialized. is this the burden of the workbench or the grid?

**Comment by sbatten** [comment]
@joaomoreno having trouble with the resizing priority feature if you could take a look

**Comment by sbatten** [comment]
@joaomoreno @bpasero this is now behind the setting `workbench.useGridLayout`. It is by default set to false and thus must be enabled to test.

**Comment by joaomoreno** [comment]
> @joaomoreno maybe the snapSize value should be a number from 0 to 1 in relation to the minimum size of the current orientation. panel needs to have a different snap value depending on panel position. also, it seems that if an item is snapped closed, that state isn't serialized. is this the burden of the workbench or the grid?

I like that snap approach very much: https://github.com/Microsoft/vscode/issues/66530

As for the serialization, it should be the grid's responsibility to serialize it properly... created this to investigate: https://github.com/Microsoft/vscode/issues/66531

> @joaomoreno having trouble with the resizing priority feature if you could take a look

What is the issue? :thinking:

**Comment by joaomoreno** [comment]
Just ran this, looks pretty cool. I've started checking off the checkboxes from my previous comment.

- Snapping serialization seems to work fine for the panel, but not for the sidebar. @sbatten Did you so something different for these?
- The pre-layout feature on reload seems to have the sizes broken. When you reload, you don't seem to get the same sidebar width on the loading ghost image at the time of reload, but the one at the time of startup.
- After toggling Zen mode, everything broke. And now the state is completely broken and I can't seem to get out of it:

image

**Comment by sbatten** [comment]
@joaomoreno @bpasero I will merge this as is after having people Selfhost a private build with the flag disabled (default).

File: src/vs/base/browser/ui/grid/grid.ts [hunk 1]
```diff
@@ -186,9 +186,10 @@ export interface IGridStyles extends IGridViewStyles { }
 
 export interface IGridOptions {
 	styles?: IGridStyles;
+	proportionalLayout?: boolean;
 }
 
-export class Grid<T extends IView> implements IDisposable {
+export class Grid<T extends IView=IView> implements IDisposable {
```
Comment (bpasero) [inline src/vs/base/browser/ui/grid/grid.ts]: @sbatten why this alias `Grid<T extends IView=IView>`?

File: src/vs/workbench/browser/part.ts [hunk 2]
```diff
@@ -22,7 +22,7 @@ export abstract class Part extends Component {
 	private parent: HTMLElement;
 	private titleArea: HTMLElement | null;
 	private contentArea: HTMLElement | null;
-	private partLayout: PartLayout;
+	protected partLayout: PartLayout;
```
Comment (bpasero) [inline src/vs/workbench/browser/part.ts]: @sbatten why is this needed? When I find references all of them seem to call `partLayout.layout(dimension)` which is exactly the same as the `layout()` function in `part.ts` does. I would suggest to leave `partLayout` private and call `super.layout(dimension)` instead.

File: src/vs/workbench/electron-browser/main.contribution.ts [hunk 3]
```diff
@@ -689,6 +689,12 @@ configurationRegistry.registerConfiguration({
 			'description': nls.localize('workbench.enableExperiments', "Fetches experiments to run from a Microsoft online service."),
 			'default': true,
 			'tags': ['usesOnlineServices']
+		},
+		'workbench.useGridLayout': {
```
Comment (bpasero) [inline src/vs/workbench/electron-browser/main.contribution.ts]: @sbatten should this have some "experimental" in its name to make clear its being removed in the future?

File: src/vs/workbench/electron-browser/workbench.ts [hunk 4]
```diff
@@ -168,8 +171,15 @@ interface IZenMode {
...
 
+interface IWorkbenchUIState {
+	lastPanelHeight?: number;
+	lastPanelWidth?: number;
+	lastSidebarDimension?: number;
+}
+
 export class Workbench extends Disposable implements IPartService {
 
+	private static readonly workbenchGridUIStateStorageKey = 'workbench.grid.state';
```
Comment (bpasero) [inline src/vs/workbench/electron-browser/workbench.ts]: @sbatten maybe better `workbench.layout.state`?

**Review by bpasero — CHANGES_REQUESTED** [review]
@sbatten I left some code review comments. On top of that:
* search for `.monaco-workbench >` and find more places where we probably need to remove the "`>`"
* running out of sources on macOS, I see the native title as well as the custom title
* `workbench.ts` seems to have inherited the entire grid layout logic making the file even harder to read, did you think about encapsulating all the layout things into a separate class, similar to `layout.ts`?

**Comment by sbatten** [comment]
I've updated the PR to address some of @bpasero's comments, I will still merge as these changes should not be breaking functionality with the flag disabled. I will use a separate issue to track additional items to address going forward. Anything not already mentioned in this PR should get its own issue and assigned to me.
